File: practices/chapter18_knn/python/ch_18_kmeans.py
import numpy as np
import matplotlib.pyplot as plt

def get_distance(x1, y1, x2, y2):
  return int(((x1 - x2) ** 2 + (y1 - y2) ** 2) ** 0.5)

def clustering(x, y, cx, cy, cluster_number, seeds_amount):
  clusters = []

  for i in range(cluster_number):
    clusters.append([]) # init
  for seed_i in range(seeds_amount):
    distance_now = np.Infinity
    cluster_index = 0
    for cluster_i in range(cluster_number):
      dist = get_distance(x[seed_i], y[seed_i], cx[cluster_i], cy[cluster_i])
      if dist < distance_now:
        distance_now = dist
        cluster_index = cluster_i
    clusters[cluster_index].append([x[seed_i], y[seed_i]]) # 將seed加入群集

  return clusters

def kmeans(x, y, cx, cy, cluster_number, seeds_amount):
  clusters = clustering(x, y, cx, cy, cluster_number, seeds_amount)
  plt.scatter(x, y, color='b')
  plt.scatter(cx, cy, color='r')

  cluster_colors = ['r', 'g', 'b']
  for index, node in enumerate(clusters):
    line_x_points = []
    line_y_points = []
    for n in node:
      line_x_points.append([n[0], cx[index]])
      line_y_points.append([n[1], cy[index]])
    color = cluster_colors[index]
    for i in range(len(line_x_points)):
      plt.plot(line_x_points[i], line_y_points[i], color=color)
  
  plt.show()
  return clusters

def make_new_cluster(clusters):
  new_x_points = []
  new_y_points = []
  for index, node in enumerate(clusters):
    node_x, node_y = 0, 0
    for n in node: # 加總該群集的每個點
      node_x += n[0]
      node_y += n[1]

    new_x_points.append([])
    new_y_points.append([])
    new_x_points[index] = int(node_x / len(node)) # 該群集的中心x
    new_y_points[index] = int(node_y / len(node)) # 該群集的中心y

  return new_x_points, new_y_points

def main():
  cluster_number = 3
  seeds = 50
  limits = 100
  
  x = np.random.randint(0, limits, seeds)
  y = np.random.randint(0, limits, seeds)
  cluster_center_x = np.random.randint(0, limits, cluster_number)
  cluster_center_y = np.random.randint(0, limits, cluster_number)
  
  init_clusters = kmeans(x, y, cluster_center_x, cluster_center_y, cluster_number, seeds)
  while True:
    new_x_points, new_y_points = make_new_cluster(init_clusters)
    x_points_list = list(cluster_center_x)
    y_points_list = list(cluster_center_y)
    if new_x_points == x_points_list and new_y_points == y_points_list:
      break
    else:
      cluster_center_x = new_x_points
      cluster_center_y = new_y_points
      init_clusters = kmeans(x, y, cluster_center_x, cluster_center_y, cluster_number, seeds)

if __name__ == "__main__":
    main()

File: practices/chapter18_knn/js/ch_18_kmeans.js
import {
  KmeansRender
} from './kmeansRender.js'

const animationDelaySecs = 1

const delayFnWrapper = (delaySecs=1) => (fn) => (...params) => (
  new Promise(res => (
    setTimeout(() => res(fn(...params)), 1000 * delaySecs)
  ))
)

function getRandom(min, max) {
  return min + Math.floor(Math.random() * (max - min))
}

function makeRandomPoints(min, max, amount) {
  let points = []
  for (let i = 0; i < amount; i++) {
    const point = [
      getRandom(min, max),
      getRandom(min, max),
    ]
    points.push(point)
  }
  return points
}

function getDistance(point1=[0, 0], point2=[0, 0]) {
  return Math.round(
    Math.pow(
      Math.pow(point1[0] - point2[0], 2) +
      Math.pow(point1[1] - point2[1], 2), 
    0.5)
  )
}

function clustering(points, clusterCenterPoints, clusterNumber, seedsAmount) {
  let clusters = [...Array(clusterNumber).keys()].map(() => [])

  for (let i = 0; i < seedsAmount; i++) {
    let distNow = Infinity
    let clusterIndex = 0
    const point = points[i]
    // 看每個點該歸類在哪個群集
    for (let j = 0; j < clusterNumber; j++) {
      const dist = getDistance(point, clusterCenterPoints[j])
      if(dist < distNow) {
        distNow = dist
        clusterIndex = j
      }
    }
    // 加入群集
    clusters[clusterIndex].push(point)
  }

  return clusters
}

async function kmeans(points=[[]], clusterCenterPoints=[[]], clusterNumber, seedsAmount) {
  const kmeansRender = new KmeansRender()
  const clusters = clustering(points, clusterCenterPoints, clusterNumber, seedsAmount)
  // render on canvas
  kmeansRender.scatter(points, '#00f')
  kmeansRender.scatter(clusterCenterPoints, '#c00')

  const clusterLineColors = ['#dd0', '#f0f', '#0cc']
  for (let i = 0; i < clusters.length; i++) {
    const cluster = clusters[i];
    const lineColor = clusterLineColors[i]
    const lineToPos = clusterCenterPoints[i]
    let clusterLines = []
    for (const pt of cluster) {
      const lineFromPos = pt
      clusterLines.push([lineFromPos, lineToPos])
    }
    kmeansRender.plot(clusterLines, lineColor)
  }

  kmeansRender.show()
  // return new Promise(res => {
  //   setTimeout(() => res(clusters), 1000 * animationDelaySecs)
  // })
  return clusters
}

function makeNewClusterCenterPoints(clusters) {
  let newClusterCenterPoints = []

  for (let i = 0; i < clusters.length; i++) {
    const cluster = clusters[i];
    let pointsSum = [0, 0]
    for (const point of cluster) {
      pointsSum[0] += point[0]
      pointsSum[1] += point[1]
    }
    newClusterCenterPoints[i] = [
      Math.round(pointsSum[0] / cluster.length),
      Math.round(pointsSum[1] / cluster.length),
    ]
  }

  return newClusterCenterPoints
}

async function getBestClusters(clusterNumber, seedsAmount, limits) {
  const points = makeRandomPoints(0, limits, seedsAmount)

  let clusterCenterPoints = makeRandomPoints(0, limits, clusterNumber)
  let clusters = await delayFnWrapper(animationDelaySecs)(kmeans)(points, clusterCenterPoints, clusterNumber, seedsAmount)

  while(true) {
    const newClusterCenterPoints = makeNewClusterCenterPoints(clusters)
    if(JSON.stringify(newClusterCenterPoints) === JSON.stringify(clusterCenterPoints)) {
      break
    } else {
      clusterCenterPoints = newClusterCenterPoints
      clusters = await delayFnWrapper(animationDelaySecs)(kmeans)(points, clusterCenterPoints, clusterNumber, seedsAmount)
    }
  }

  return clusters
}

async function main() {
  const clusterNumber = 3
  const seeds = 100
  const limits = 400

  const res = await getBestClusters(clusterNumber, seeds, limits)
  console.log(res)
}

function setup() {
  main()
  const genKmeansButton = document.getElementById('genKmeans')
  genKmeansButton.addEventListener('click', () => {
    main()
  })
}

setup()